packages/components/src/QueryModel/ChartMenu.tsx

```diff
@@ -0,0 +1,104 @@
+import React, { PureComponent, ReactNode } from 'react';
+import { DropdownButton, MenuItem } from 'react-bootstrap';
+
+import { DataViewInfoTypes } from '../constants';
+import { ChartMenuItem } from '../components/chart/ChartMenuItem';
+import { DataViewInfo } from '../models';
+
+import { ChartModal } from '../components/chart/ChartModal';
+import { blurActiveElement } from '../util/utils';
+
+import { RequiresModelAndActions } from './withQueryModels';
+
+interface Props extends RequiresModelAndActions {
+    onChartClicked?: (chart: DataViewInfo) => boolean;
+    onCreateReportClicked?: (type: DataViewInfoTypes) => void;
+    showSampleComparisonReports: boolean;
+}
+
+interface State {
+    selectedChart?: DataViewInfo;
+}
+
+export class ChartMenu extends PureComponent<Props, State> {
+    constructor(props) {
+        super(props);
+
+        this.state = { selectedChart: undefined };
+    }
+    componentDidMount(): void {
+        const { model, actions, showSampleComparisonReports } = this.props;
+        actions.loadCharts(model.id, showSampleComparisonReports);
+    }
+
+    onCreateClicked = (): void => {
+        // We only support creating SampleComparison reports at the moment.
+        this.props.onCreateReportClicked(DataViewInfoTypes.SampleComparison);
+    };
+
+    chartClicked = (chart): void => {
+        const { onChartClicked } = this.props;
+
+        blurActiveElement();
+
+        // If the user supplies a click handler then we use the response from that to determine if we should render the
+        // chart modal. This is needed so Biologics and redirect to Sample Comparison Reports.
+        if (onChartClicked === undefined || onChartClicked(chart)) {
+            this.setState({ selectedChart: chart });
+        }
+    };
+
+    clearChart = (): void => {
+        this.setState({ selectedChart: undefined });
+    };
+
+    render(): ReactNode {
+        const { model, showSampleComparisonReports } = this.props;
+        const { charts, chartsError, hasCharts, id, isLoading, isLoadingCharts } = model;
+        const privateCharts = hasCharts ? charts.filter(chart => !chart.shared) : [];
+        const publicCharts = hasCharts ? charts.filter(chart => chart.shared) : [];
+        const title = isLoadingCharts ? <span className="fa fa-spinner fa-pulse" /> : 'Charts';
+        const chartMapper = (chart): ReactNode => (
+            <ChartMenuItem key={chart.reportId} chart={chart} showChart={this.chartClicked} />
+        );
+        const noCharts = hasCharts && charts.length === 0 && !showSampleComparisonReports;
+        const disabled = isLoading || isLoadingCharts || noCharts;
+        const { selectedChart } = this.state;
+        // Have to safe guard around isLoading here because model.filters getter will throw an error if there is no
```

Comment:
Yeah I think I am going to make model.filters not throw an error, I'm going to move that error to the QueryModelLoader, but I also don't like the behavior from QueryGridModel, we should chat before I merge, I have an idea.

Reply:
What I ended up doing was moving the code around so I could use filters without triggering the error, which I think is correct. No one should be accessing the filters attribute until we can properly construct the filters array. If we're going to return anything instead of raising an exception it should be `undefined`. An empty array filter is a valid state for the model to be in, however, before we have a QueryInfo available to us returning an empty array would be incorrect, but you would not be able to tell that unless you also check the status of QueryInfo, so I'd rather just return undefined if we're going to choose not to throw an error. For the time being I am going to defer making any change as we no longer have any use cases where throwing an error is inconvenient or incorrect.